### PR TITLE
[management] smoke test!

### PR DESCRIPTION
**Pull request by davidiw**
Per title, this introduces an end-to-end smoke test for management.
The test completes because all nodes wre able to get connected.
this shows that there are still a few missing pieces:
* waypoint generation / insertion
* making sense of fullnodes
* strip out network signing key or evaluate if we want it there

some cleanups:
* a better framework for converting ed25519 to x25519 so we don't need
to export the ed25519 private key
* better integration or replacement of libra swarm

### CODE REVIEW

**Comment by huitseeker** [comment]
See also #3904

File: consensus/safety-rules/src/safety_rules.rs [hunk 1]
```diff
@@ -120,7 +120,7 @@ impl<T: Payload> SafetyRules<T> {
                 .set_waypoint(&Waypoint::new_epoch_boundary(ledger_info)?)?;
             self.persistent_storage.set_last_voted_round(0)?;
             self.persistent_storage.set_preferred_round(0)?;
-            self.persistent_storage.set_epoch(ledger_info.epoch() + 1)?;
+            self.persistent_storage.set_epoch(next_epoch)?;
```
Comment (zekun000) [inline consensus/safety-rules/src/safety_rules.rs:123]: you can use epoch_info.epoch and skip the next_epoch var

File: config/management/src/smoke_test.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            .unwrap();
+        let identity_key = identity_key.to_bytes().as_ref().try_into().unwrap();
+        // This is a deprecated field
+        let signing_key = Ed25519PrivateKey::generate_for_testing();
+        fullnode_network.network_keypairs = Some(NetworkKeyPairs::load(identity_key, signing_key));
+        let fullnode_network_address = fullnode_network.listen_address.clone();
+
+        configs.push(config);
+
+        helper.operator_key(&ns, &ns_shared).unwrap();
```
Comment (zekun000) [inline config/management/src/smoke_test.rs:100]: those helper functions are so hard to read, spend a few mins realizing this is to fetch from local and submit to remote

File: config/management/src/smoke_test.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    for i in 0..num_validators {
+        let storage = helper.storage(i.to_string());
+        let ns = i.to_string();
+        let ns_shared = ns.clone() + shared;
+        helper.initialize(ns.clone());
+
+        let mut config = NodeConfig::default();
+
+        let mut network = NetworkConfig::default();
+        network.discovery_method = DiscoveryMethod::Onchain;
```
Comment (zekun000) [inline config/management/src/smoke_test.rs:70]: milestone!

File: config/management/src/smoke_test.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    for (i, mut config) in configs.iter_mut().enumerate() {
+        let waypoint_value = Value::String(waypoint.to_string());
+        helper
+            .storage(i.to_string())
+            .set(libra_global_constants::WAYPOINT, waypoint_value)
+            .unwrap();
+
+        let mut to = swarm_path.clone();
+        to.push("secure_store_for_".to_string() + &i.to_string());
```
Comment (zekun000) [inline config/management/src/smoke_test.rs:136]: use format?

File: config/management/src/smoke_test.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        std::fs::copy(helper.path_string(), &to).unwrap();
+
+        let mut storage_config = OnDiskStorageConfig::default();
+        storage_config.path = to;
+        storage_config.set_data_dir(PathBuf::from(""));
```
Comment (zekun000) [inline config/management/src/smoke_test.rs:141]: is this not used?

File: config/management/src/smoke_test.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    // Step 3) Prepare validators
+    let mut configs = Vec::new();
+    for i in 0..num_validators {
+        let storage = helper.storage(i.to_string());
+        let ns = i.to_string();
+        let ns_shared = ns.clone() + shared;
+        helper.initialize(ns.clone());
+
+        let mut config = NodeConfig::default();
+
```
[the remark this answers is not in the document]
Comment (davidiw) [reply config/management/src/smoke_test.rs:68]: it is a separate clause from network and has nothing to do with ns initialization

File: config/management/src/smoke_test.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        let mut network = NetworkConfig::default();
+        network.discovery_method = DiscoveryMethod::Onchain;
+        config.validator_network = Some(network);
+
+        let mut network = NetworkConfig::default();
+        network.discovery_method = DiscoveryMethod::Onchain;
+        config.full_node_networks = vec![NetworkConfig::default()];
```
Comment (davidiw) [inline config/management/src/smoke_test.rs:75]: woops, good call, I added fullnode support but don't use it, so it is untested... I'll clean this up

File: config/management/src/smoke_test.rs (hunk 2 above)
Comment (davidiw) [reply config/management/src/smoke_test.rs:100]: ahh, the intent is to ensure the command-line works as possible, I guess I could add comments to storage_helper... thoughts on what the confusion stemmed from?

File: config/management/src/smoke_test.rs (hunk 5 above)
Comment (davidiw) [reply config/management/src/smoke_test.rs:141]: it defaults to something, but no, it is not, if the path is a full path, this is ignored, but I want to be careful to not leave some unused data here.

File: consensus/safety-rules/src/safety_rules.rs (hunk 1 above)
Comment (davidiw) [reply consensus/safety-rules/src/safety_rules.rs:123]: oh, this must have appeared after your code refactor, it wasn't available before :)

File: config/management/src/smoke_test.rs (hunk 2 above)
Comment (zekun000) [reply config/management/src/smoke_test.rs:100]: I'd expect something like `submit_[operator_key|validator_config]`, `operator_key` feels the function only returns the operator key without doing anything else

**Pull request closed without merging**